### PR TITLE
[rubysrc2cpg] kick-starting the new AstCreator

### PR DESCRIPTION
**Pull request by xavierpinho**
As a follow-up for #3755 and #3764, this brings in the current work-in-progress AstCreator that uses the new RubyParser.g4

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  protected val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  protected val relativeFileName: String =
+    projectRoot.map(filename.stripPrefix).map(_.stripPrefix(java.io.File.separator)).getOrElse(filename)
+
+  override def createAst(): BatchedUpdate.DiffGraphBuilder = {
+    parser.parse(filename) match
+      case Success(programCtx) =>
+        // TODO:
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:33]: What's the TODO here?

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  protected def astForFieldAccess(node: MemberAccess): Ast = {
+    val fieldIdentifierAst = Ast(fieldIdentifierNode(node, node.methodName, node.methodName))
+    val targetAst          = astForExpression(node.target)
+    val code               = s"${node.target.getText}${node.op}${node.methodName}"
+    val fieldAccess = callNode(node, code, Operators.fieldAccess, Operators.fieldAccess, DispatchTypes.STATIC_DISPATCH)
+    callAst(fieldAccess, Seq(targetAst, fieldIdentifierAst))
+  }
+
+  private def getBinaryOperatorName(op: String): Option[String] = {
+    Map(
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala:257]: This Map could be defined statically somewhere instead of creating it here with every call to `getBinaryOperatorName`.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+      "&"   -> Operators.and,
+      "|"   -> Operators.or,
+      "^"   -> Operators.xor,
+      "<<"  -> Operators.shiftLeft,
+      ">>"  -> Operators.logicalShiftRight
+    ).get(op)
+  }
+
+  private def getUnaryOperatorName(op: String): Option[String] = {
+    Map(
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala:284]: See comment above.

**Review by max-leuthaeuser — APPROVED** [review]
Just a few minor things. 
Otherwise, lgtm.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    methodAstParentStack.push(namespaceBlock)
+    scope.pushNewScope(namespaceBlock)
+    val rubyFileMethod = astInFakeMethod(rootStatements)
+    scope.popScope()
+    methodAstParentStack.pop()
+    Ast(fileNode).withChild(Ast(namespaceBlock).withChild(rubyFileMethod))
+  }
+
+  private def astInFakeMethod(rootNode: StatementList): Ast = {
+    val name     = ":program" // TODO: avoid this hardcoding. Move it into Defines?
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:63]: > `// TODO: avoid this hardcoding. Move it into Defines?`

Yeah, it's used across multiple passes, I would agree to doing this. Though it can be its own PR.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+// TODO: Move this to a more appropriate place?
+object GlobalTypes {
+  val builtinPrefix = "__builtin"
+
+  /* Sources:
+   * https://ruby-doc.org/docs/ruby-doc-bundle/Manual/man-1.4/function.html
+   * https://ruby-doc.org/3.2.2/Kernel.html
+   *
+   * We comment-out methods that require an explicit "receiver" (target of member access.)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala:44]: > We comment-out methods that require an explicit "receiver" (target of member access.)

Are these inherited by some top-level object? Like Java's `java.lang.Object`? Might be useful to stub these, though just a thought for future.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala (hunk 1 above)
Comment (xavierpinho) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:33]: Thanks for catching it! It was a leftover. Will remove it.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  final case class SimpleIdentifier(ctx: ParserRuleContext, typeFullName: Option[String] = None) extends ParserNode(ctx)
+
+  final case class SelfIdentifier(ctx: ParserRuleContext) extends ParserNode(ctx)
+
+  /** Represents a non-interpolated literal. */
+  final case class StaticLiteral(ctx: ParserRuleContext, typeFullName: String) extends ParserNode(ctx) {
+    def isSymbol: Boolean = ctx.getText.startsWith(":")
+    def isString: Boolean = ctx.getText.startsWith("\"")
+
+    // TODO: ugly as hell.
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala:136]: Lol, the other solution I can think of is pattern matching, i.e.,:
```scala
  def innerText: String = ctx.getText.toList match {
    case ':' :: '\'' :: tail => tail.dropRight(1)
    case ':' :: tail         => tail
    case '\'' :: tail        => tail.dropRight(1)
    case _                   => ctx.getText
  }
```

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
+  final case class DynamicLiteral(ctx: ParserRuleContext, typeFullName: String, expressions: List[ParserRuleContext])
+      extends ParserNode(ctx)
+
+  final case class RangeExpression(ctx: ParserRuleContext, lowerBound: ParserRuleContext, upperBound: ParserRuleContext)
+      extends ParserNode(ctx)
+
+  final case class ArrayLiteral(ctx: ParserRuleContext, elements: List[ParserRuleContext]) extends ParserNode(ctx) {
+    // TODO: ugly as hell.
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala:158]: > // TODO: ugly as hell.

Probably would pattern match as above but maybe overkill for the ones below

**Review by DavidBakerEffendi — APPROVED** [review]
Looks great! Looking forward to this

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala (hunk 6 above)
Comment (maltek) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala:136]: it's also possible to pattern-match on s-strings, i.e.
```scala
case s":'$content'" => content
case s":$symbol" => symbol
case s"'$content'" => content
case _ => ctx.getText
```

Comment (xavierpinho) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala:136]: I should probably pattern match on `ctx` instead, but that would require that I rearrange some ANTLR rules beforehand and it just didn't seem that high priority at the moment. But yeah, this hurts to look at. Thanks for the suggestion as well :)

Comment (xavierpinho) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/parser/ParserAst.scala:136]: > it's also possible to pattern-match on s-strings

Cool! I didn't know that. Thanks for mentioning it!
